TT bootloader style update

TychoVrahe opened the pull request:
Had to adjust backlight PWM setting a bit, to make transitions bootloader->firmware look good without flickering.

Cherrypicked fonts from https://github.com/trezor/trezor-firmware/pull/2876 for now.

The RSOD screens still use paragraphs, which i need for vertical centering that Label cannot do yet - but i believe that such functionality would be useful for Label as well a would simplify things here a bit.

Handles "uninitialized" boot flow as well. 

Closes #2866 
Closes #2823

grdddj commented on core/embed/rust/src/ui/model_tt/bootloader/confirm.rs:
```diff
@@ -98,17 +98,49 @@ impl<'a> Component for Confirm<'a> {
+            40 + CONTENT_PADDING
         } else {
-            0
+            58
```
These numbers could be constants

TychoVrahe replied:
https://github.com/trezor/trezor-firmware/commit/ec0f08b337177b98bd9cae8aedae48ebbaab2618

grdddj commented on core/embed/rust/src/ui/model_tt/bootloader/confirm.rs:
```diff
@@ -98,17 +98,49 @@ impl<'a> Component for Confirm<'a> {
 
-        let button_size = Offset::new(106, 38);
+        let content_area = Rect::new(
+            Point::new(CONTENT_PADDING, content_area_start), //66 if icon instead of
```
Outdated comment?

TychoVrahe replied:
https://github.com/trezor/trezor-firmware/commit/ec0f08b337177b98bd9cae8aedae48ebbaab2618

grdddj commented on core/embed/rust/src/ui/model_tt/bootloader/confirm.rs:
```diff
@@ -175,11 +209,12 @@ impl<'a> Component for Confirm<'a> {
             if let Some(icon) = self.icon {
                 icon.draw(
-                    Point::new(screen().center().x, 32),
+                    Point::new(screen().center().x, 17),
```
Unnamed constant

TychoVrahe replied:
https://github.com/trezor/trezor-firmware/commit/ec0f08b337177b98bd9cae8aedae48ebbaab2618

grdddj commented on core/embed/rust/src/ui/model_tt/bootloader/mod.rs:
```diff
@@ -511,3 +531,158 @@ extern "C" fn screen_welcome() {
+        unsafe { from_c_str(version) }
+    };
+
+    fadeout();
```
Any reason to do `fadeout()` here in the middle of assigning? I would prefer all the actions be together, so they are better visible - in this case above `display::rect_fill` probably

TychoVrahe replied:
https://github.com/trezor/trezor-firmware/commit/35391d98c6e13fa84c443ec624cd949e7964e1db

grdddj commented on core/embed/rust/src/ui/model_tt/bootloader/mod.rs:
```diff
@@ -511,3 +531,158 @@ extern "C" fn screen_welcome() {
+        }
+    }
+
+    if (vtrust & (VTRUST_CLICK as u16)) == 0 {
```
What is this `vtrust` thing?

TychoVrahe replied:
its from vendor header - contains flags about how should the firmware be trusted. e.g. firmware signed by satoshi labs is trusted and doesn't require user confirmation when starting up, whereas development firmware does.

grdddj commented on core/embed/rust/src/ui/model_tt/component/error.rs:
```diff
@@ -0,0 +1,104 @@
+
+pub struct ErrorScreen<T> {
+    bg: Pad,
+    small_pad: Pad,
```
Could this be called `footer_pad`? It is being placed on the same area

TychoVrahe replied:
removed, see below.

https://github.com/trezor/trezor-firmware/commit/4acf54ce632927f4117f1491055323fbbc73c50b

in result.rs i renamed it: https://github.com/trezor/trezor-firmware/commit/7397275c9d65c0946e0c7c6fc70a0f88194d3282

grdddj commented on core/embed/rust/src/ui/screens.rs:
```diff
@@ -8,33 +8,26 @@ pub use super::model_tt::screens::*;
-        unsafe { from_c_str(msg) }
-    };
-    let file = if file.is_null() {
+extern "C" fn screen_fatal_error_r(
```
What is the meaning of `_r` suffix? If it is `Rust`, I would write `_rust` as my first thought was `model R`

TychoVrahe replied:
https://github.com/trezor/trezor-firmware/commit/fafa8e4bd46f5f04e2266ae88570675357b4a695

note that in `rust_ui.h` there is another similarly named function `loader_uncompress_r`, i left it like that for now as it will be hopefully removed entirely in near future.

grdddj commented on core/embed/rust/src/ui/screens.rs:
```diff
@@ -8,33 +8,26 @@ pub use super::model_tt::screens::*;
+    msg: *const cty::c_char,
+    footer: *const cty::c_char,
+) {
+    let title = if title.is_null() {
```
We use this construct three times in a row here, would it be worth to create a macro out of this?

```rs
if XXX.is_null() {
    ""
} else {
    unwrap!(unsafe { from_c_str(XXX) })
}
```

TychoVrahe replied:
https://github.com/trezor/trezor-firmware/commit/72eb69847e26b134df18f21ad7c7169b0a7c74c8

grdddj reviewed (commented):
Some comments, without trying to run the code - it probably needs a devkit.

Overall, there are quite a lot of magic numbers that we might want to extract into constants

TychoVrahe commented on core/embed/rust/src/ui/model_tt/bootloader/confirm.rs:
```diff
@@ -48,38 +47,39 @@ impl<'a> Confirm<'a> {
         icon: Option<Icon>,
         left: Button<&'static str>,
```
https://github.com/trezor/trezor-firmware/commit/ec0f08b337177b98bd9cae8aedae48ebbaab2618

TychoVrahe commented on core/embed/rust/src/ui/model_tt/bootloader/menu.rs:
```diff
@@ -64,14 +59,19 @@ impl Component for Menu {
-            Point::new(CONTENT_PADDING, 64),
-            Point::new(WIDTH - CONTENT_PADDING, 64 + 38),
+            Point::new(CONTENT_PADDING, 56),
+            Point::new(WIDTH - CONTENT_PADDING, 56 + BUTTON_HEIGHT),
```
https://github.com/trezor/trezor-firmware/commit/ec0f08b337177b98bd9cae8aedae48ebbaab2618

TychoVrahe commented on core/embed/rust/src/ui/model_tt/bootloader/mod.rs:
```diff
@@ -493,6 +505,14 @@ extern "C" fn screen_install_success(
+#[no_mangle]
+extern "C" fn screen_welcome_model() {
+    let mut frame = WelcomeScreen::new();
+    frame.place(screen());
```
https://github.com/trezor/trezor-firmware/commit/8aa8e6ffe92bb9f588499739315c7fcc882df09e

TychoVrahe commented on core/embed/rust/src/ui/model_tt/bootloader/mod.rs:
```diff
@@ -511,3 +531,158 @@ extern "C" fn screen_welcome() {
+            while wait > 0 {
+                let mut buff = [0u8; 10];
+                let mut wait_str: String<32> = String::new();
+                wait_str.push_str("starting in ").unwrap();
```
https://github.com/trezor/trezor-firmware/commit/1aa2b68a9b0d6da8d16b45b238aca94f1ad32886

TychoVrahe commented on core/embed/rust/src/ui/model_tt/component/error.rs:
```diff
@@ -0,0 +1,104 @@
+
+    fn paint(&mut self) {
+        self.bg.paint();
+        self.small_pad.paint();
```
yes its a remnant from result.rs, removed: https://github.com/trezor/trezor-firmware/commit/4acf54ce632927f4117f1491055323fbbc73c50b

TychoVrahe commented on core/embed/rust/src/ui/model_tt/component/error.rs:
```diff
@@ -0,0 +1,104 @@
+        let mut instance = Self {
+            bg: Pad::with_background(FATAL_ERROR_COLOR),
+            small_pad: Pad::with_background(FATAL_ERROR_COLOR),
+            title: Label::new(title, Center, TEXT_ERROR_BOLD).into_child(),
```
https://github.com/trezor/trezor-firmware/commit/323c92019d1b13636918ecfc15990236119c4264

TychoVrahe replied:
https://github.com/trezor/trezor-firmware/commit/f98f9a595e719d35637a55db5c37bfe7107d7405

TychoVrahe replied:
and https://github.com/trezor/trezor-firmware/commit/cc5233e573ea92e5748df12b87e6b1a5d48e2ab7

TychoVrahe commented on core/embed/rust/src/ui/model_tt/component/result.rs:
```diff
@@ -1,45 +1,49 @@
+
+const PADDING: i16 = 6;
+const FOOTER_START: i16 = 178;
+const FOOTER_HEIGHT: i16 = 56;
```
https://github.com/trezor/trezor-firmware/commit/323c92019d1b13636918ecfc15990236119c4264

TychoVrahe commented on core/embed/trezorhal/displays/st7789v.h:
```diff
@@ -21,6 +21,6 @@ extern __IO uint8_t *const DISPLAY_DATA_ADDRESS;
 void display_set_little_endian(void);
 void display_set_big_endian(void);
-void display_set_slow_pwm(void);
+void display_set_slow_pwm();
```
https://github.com/trezor/trezor-firmware/commit/653db65ab9448bd769cd48432db76a1ce9c8e6a7

TychoVrahe commented on core/embed/rust/src/ui/model_tt/bootloader/confirm.rs:
```diff
@@ -48,38 +47,39 @@ impl<'a> Confirm<'a> {
         right: Button<&'static str>,
-        confirm_left: bool,
-        confirm: (Option<&'static str>, Paragraphs<ParagraphVecShort<&'a str>>),
+        confirm: (
```
agree about readability, but i kinda want to avoid having to split the `Confirm` into something like `ConfirmWipe` and `ConfirmInstall`, and if i just split the `confirm` argument `clippy` hates me for using 8 arguments. any ideas how to simplify here?

grdddj replied:
I did not know about this `many arguments` limit of `clippy`. In that case, this approach is fine, I cannot suggest something quick and easy. It should be possible to somehow refactor and simplify the whole `Confirm` struct, but this is a `bootloader` code that will not be updated frequently, so IMO it is fine as is

grdddj commented on core/embed/rust/src/ui/model_tt/bootloader/intro.rs:
```diff
@@ -44,11 +47,11 @@ impl<'a> Intro<'a> {
             bg: Pad::with_background(BLD_BG),
```
Most of the time we are using little different `with_...` method appearance:

```rs
pub fn with_clear(mut self) -> Self {
    self.clear = true;
    self
}
```
instead of 
```rs
pub fn with_clear(self) -> Self {
    Self {
        clear: true,
        ..self
    }
}
```

matejcik replied:
yeah but this one is better :) reason being, the latter can be `const` while the former can't

grdddj replied:
Does it mean that it would be beneficial for the codesize to use it in all the places and make them all `const`?

matejcik replied:
You can try. But the last time I tried this the saving was marginal, on the order of single bytes.

This makes sense when you need a const value for something (and so your `new()` is also const), otherwise it's just a matter of style IMO.

grdddj reviewed (approved):
I think it is alright

TychoVrahe commented:
Pushed two more commits after going over this with @Hannsek 
- adjusted title position to be better centered vertically, fixed vertical positioning of "FINGERPRINT" title
- changed font of welcome URL to normal
- changed forced delay at startup to (while showing model name) to 1200ms. 
- faster fading when changing bootloader screens

Pending changes:
- change font of model name to DEMIBOLD, will be in bootloader done by using icon, waiting for pixel perfect png.
- check text appearance after rebasing onto firmware style updare - font spacing will affect bootloader too

matejcik commented:
added the bootloader emulator.

build with `make build_bootloader_emu`
run with `./build/bootloader_emu/bootloader.elf`
click on the lock to enter bootloader, or run `bootloader.elf s` to go directly in

Hannsek commented:
I don’t remember such design. Probably misunderstanding between us. 🤷🏻‍♂️

But it is not needed.

TychoVrahe commented:
Rebased. 

Fixed few issues with the emulator - style and compatibility with latest changes in master. 

Removed the darker boot logo.

matejcik commented:
rebased again to fix conflicts

reverted "boot click in rust"

added some minor tweaks; most changes that I originally made were in the "boot click" part of code so :)

TychoVrahe reviewed (commented):
Reviewed matejciks commits.

Appreaciate the simplifications made by removing paragraphs, vertically aligned label etc, helps a lot.

I would prefer the Label to be using Alignment2D for vertical alignment like icons, but since this orignates from previous PR lets keep it this way.

The bootloader emulator deserves a mention it docs, lets make at least an issue for that.

Hannsek commented:
> The bootloader emulator deserves a mention it docs, lets make at least an issue for that.

https://github.com/trezor/trezor-firmware/issues/2926

matejcik commented:
fixed the seed erasing bug in af6c89f2c59f0c2b390e3f3a7fa7a7c7c827e275 hopefully

matejcik commented:
squashed & rebased on master. 
one big style-only change is cb4db82, error strings are tweaked in [c513cc8](https://github.com/trezor/trezor-firmware/pull/2879/commits/c513cc8615a94d437b0634c7b7a438806923e404), otherwise this should be the same as previous code.

QA bootloader fingerprint is `dc668381ab9e0980f3c1675d384556209c7c5bea5d712df3c06ce0fe63d13f7b`
production bootloader fingerprint is `6e31559f12f072025d151362216663013b251d7858d3ce72c1f5ee225e58447f`
@TychoVrahe please confirm

matejcik commented:
of course there are last-moment changes.

i'm going to rename [02a9519](https://github.com/trezor/trezor-firmware/pull/2879/commits/02a9519d1f7df85d960ec0bae82f0a622bef8db6) so that it's a separate commit, and tag the bootloader one commit before that.
so the only change that touches bootloader is [a9f1570](https://github.com/trezor/trezor-firmware/pull/2879/commits/a9f1570d660da9212ec8d93643510519a06128be) which I'm going to risk passing without a separate QA round.

matejcik commented:
QA bootloader fingerprint: `8af2955cad3def2357704638dfd8b4e7a8d6bf04b53e3cb6a005fa6c5ecc0d62`
production bootloader fingerprint: `a75f07d558bfd2a882e98e51a8e645c0cccc05dda9b301d6d293dbc8bbdfacf7`

tagged [9baf52e](https://github.com/trezor/trezor-firmware/pull/2879/commits/9baf52eedce5855d461e4adc5e9cb3730855ee25) as `core/bl2.1.0` so these fingerprints are for that tag, please check against it

matejcik commented:
production fingerprint as of tag `core/bl2.1.0`: `35fbb85d27601a474ffb0095ccf88de089b794f1ddc419a112100d36c2cd0f64`
(final version, hopefully.)
